### PR TITLE
[NativeAOT-LLVM]: Struct padding on the shadow stack

### PR DESCRIPTION
**Pull request by yowl**
This PR adds a call back to calculate the padding required for structs on the shadowstack, and removes the forced failures for this scenario.  Also 

- Allows multiple returns
- Changes `OperIsBlkOp ` to explicitly check the operands instead of `Data()  != nullptr` <-  Shall I create a PR to push that upstream?
- For Blk operations sets the layout and `gtBlkOpKind`. I fixed it to  `BlkOpKindHelper`  which is just a guess.

### CODE REVIEW

File: src/coreclr/jit/gentree.h
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.h]:
```suggestion
    return ((gtOper == GT_ASG) && varTypeIsStruct(AsOp()->gtOp1)) || OperIs(GT_STORE_DYN_BLK, GT_STORE_BLK, GT_STORE_OBJ);
```
Since `OperIsBlk` just checks for these same opers.

>  Shall I create a PR to push that upstream?

Not-a-maintainer's-opinion: yes.

File: src/coreclr/jit/llvm.cpp
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.cpp]: ...will always be zero as the outgoing args size is calculated during simple lowering.

I think we can just ignore it.

File: src/coreclr/jit/llvm.cpp
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.cpp]: What is the reason for using `lvClassHnd`? It is only used for type optimizations in the compiler and I suspect is not available in a number of cases. I would have expected we would use `GetStructHnd` here.

File: src/coreclr/jit/llvm.cpp
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.cpp]: I would say leave this as `BlkOpKindInvalid` - we won't be using it.

Normally it is set during lowering so that the register allocator knows how many temporary registers are needed for the copy, but we'll have LLVM figure this out by itself.

File: src/coreclr/jit/llvm.cpp
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.cpp]: Ditto: `lvClassHnd`. Is there a specific need to pass this down?

File: src/coreclr/jit/llvm.h
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.h:172]: General question: do you expect we'll leave this "in the VM" in the long term or is this a temporary solution to have the IL/RyuJit - codegens agree with each other? I ask because if we think we should leave this logic to the VM, we should probably think of ways to make transmitting the results cheaper (RPI transitions in loops seem expensive).

File: src/coreclr/jit/llvm.cpp
Comment (yowl) [reply src/coreclr/jit/llvm.cpp]: thanks

File: src/coreclr/jit/llvm.cpp
Comment (yowl) [reply src/coreclr/jit/llvm.cpp]: Right, thanks

File: src/coreclr/jit/llvm.cpp
Comment (yowl) [reply src/coreclr/jit/llvm.cpp]: I'm changing this to be clearer I hope, it will just have the struct class handle.

File: src/coreclr/jit/llvm.h
Comment (yowl) [reply src/coreclr/jit/llvm.h:172]: We'll need it long term I think.  The VM defines the layout I believe and we need the alignment and packing size from that.  If we dont use the same layout then I think structs in structs for example, would be a problem.

**Comment by yowl** [comment]
I tidied this up addressing the feedback.  I removed the `spilledExpression` vector, there was nothing in it at present.  It can be put back when/if  required.

File: src/coreclr/jit/llvm.cpp
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.cpp:383]:
```suggestion
    return varTypeIsStruct(varDsc) ? varDsc->GetStructHnd() : NO_CLASS_HANDLE;
```

**Review by SingleAccretion — APPROVED** [review]
LGTM, modulo one nit - feel free to address separately (to not spin up the CI again).

File: src/coreclr/jit/llvm.cpp
Comment (yowl) [reply src/coreclr/jit/llvm.cpp:383]: Thanks, I'll get that in the next PR.

**Comment by yowl** [comment]
@jkotas   I'll get the last suggestion in the next PR.  Pending comments from yourself and others this is ok to go.

**Review by jkotas — APPROVED** [review]
@yowl @SingleAccretion It is a pleasure to see you working together on this!

**Pull request merged**
